ontolearner/learner/text2onto/sbunlp.py

```diff
@@ -0,0 +1,525 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import random
+import re
+import ast
+import gc
+from typing import Any, Dict, List, Optional, Set, Tuple
+from collections import defaultdict
+
+import torch
+from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
+
+from ...base import AutoLearner, AutoLLM
+
+# -----------------------------------------------------------------------------
+# Concrete AutoLLM: local HF wrapper that follows the AutoLLM interface
+# -----------------------------------------------------------------------------
+class LocalAutoLLM(AutoLLM):
+    """
+    Handles loading and generation for a Hugging Face Causal Language Model (Qwen/TinyLlama).
+    Uses 4-bit quantization for efficiency and greedy decoding by default.
+    """
+
+    def __init__(self, label_mapper: Any = None, device: str = "cpu", token: str = "") -> None:
+        super().__init__(label_mapper=label_mapper, device=device, token=token)
+        self.model = None
+        self.tokenizer = None
+
+    def load(self, model_id: str, load_in_4bit: bool = False, dtype: str = "auto", trust_remote_code: bool = True):
+        """Load tokenizer + model, applying 4-bit quantization if specified and possible."""
+
+        # Determine the target data type (default to float32 for CPU, float16 for GPU)
+        torch_dtype_val = (torch.float16 if torch.cuda.is_available() else torch.float32)
+
+        # Load the tokenizer
+        self.tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=trust_remote_code)
+        if self.tokenizer.pad_token is None:
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        quant_config = None
+        if load_in_4bit:
+            # Configure BitsAndBytes for 4-bit loading
+            quant_config = BitsAndBytesConfig(
+                load_in_4bit=True,
+                bnb_4bit_compute_dtype=torch.float16,
+                bnb_4bit_use_double_quant=True,
+                bnb_4bit_quant_type="nf4",
+            )
+            if torch_dtype_val is None:
+                torch_dtype_val = torch.float16
+
+        # Set device mapping (auto for multi-GPU or single GPU, explicit CPU otherwise)
+        device_map = "auto" if (self.device != "cpu") else {"": "cpu"}
+
+        # Load the Causal Language Model
+        self.model = AutoModelForCausalLM.from_pretrained(
+            model_id,
+            device_map=device_map,
+            torch_dtype=torch_dtype_val,
+            quantization_config=quant_config,
+            trust_remote_code=trust_remote_code,
+        )
+
+        # Ensure model is on the correct device (redundant if device_map="auto" but safe)
+        if self.device == "cpu":
+            self.model.to("cpu")
+
+    def generate(self, inputs: List[str], max_new_tokens: int = 64, temperature: float = 0.0, top_p: float = 1.0) -> List[str]:
```

Comment:
follow the AutoLLM generate argument pattern! don't add args that are not there in this function args, but inside the __init__ also it would be great to leave temrature and top_p as a default.